### PR TITLE
Slower Online -> Target

### PR DESCRIPTION
**Pull request by sandhawalia**
1. Add more augmentations
2. Save patches and augmentation for debugging
3. Run validation to see if the world is collapsing

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -11,6 +11,7 @@ einops = "^0.3.0"
 torch = "^1.7.0"
 torchvision = "^0.8.1"
 tqdm = "^4.50.2"
+scikit-image = "^0.17.2"
```
Comment (daniel-j-h) [inline pyproject.toml]: We have PIL already! :hugs:

File: docker-compose.override.yaml [hunk 2]
```diff
@@ -12,3 +12,4 @@ services:
       - ./.flake8:/home/python/app/.flake8
       - ./visiont:/home/python/app/visiont
       - /tmp:/data
+      - /nas:/nas
```
Comment (daniel-j-h) [inline docker-compose.override.yaml:15]: This and the line above are a bit ugly, because it will only work on our machines :see_no_evil:

File: visiont/models.py [hunk 3]
```diff
@@ -26,22 +26,22 @@ def __init__(self, num_classes, C=3, H=224, W=224, P=16):
         self.class_token = nn.Parameter(torch.rand(D))
 
         # Standard transformer encoder, baseline defaults
-        self.enc = nn.TransformerEncoder(num_layers=12,
-                encoder_layer=nn.TransformerEncoderLayer(D, nhead=12, dim_feedforward=3072, activation="gelu"))
+        self.enc = nn.TransformerEncoder(
+            num_layers=6,
```
Comment (daniel-j-h) [inline visiont/models.py:30]: Why the change to 6 here? Just to experiment?

Comment (daniel-j-h) [reply visiont/models.py:30]: Same with heads below

Comment (daniel-j-h) [reply visiont/models.py:30]: And feedforward

Comment (sandhawalia) [reply visiont/models.py:30]: yeah iterate faster for now. Else its slow and batch size is small

File: pyproject.toml (hunk 1 above)
Comment (sandhawalia) [reply pyproject.toml]: Da next chageset would have it

File: visiont/models.py [hunk 4]
```diff
@@ -26,22 +26,22 @@ def __init__(self, num_classes, C=3, H=224, W=224, P=16):
...
+        )
 
-
-    @torch.cuda.amp.autocast()
     def forward(self, x):
         # x is N-sized batch of M P-sized patches; splitting images into
         # patches must happen in transforms to not block the mainloop
-
-        N, M, P = x.size()
+        N, T, D = x.size()
```
Comment (daniel-j-h) [inline visiont/models.py:44]: Why T?

File: visiont/scripts/__main__.py [hunk 5]
```diff
@@ -12,10 +13,66 @@ def main():
...
+    )
+    data_gen = subcmd.add_parser(
+        "generate",
+        help="generates data to see training sample diversity",
+        formatter_class=Formatter,
+    )
+    train.add_argument(
+        "-d",
+        "--dataset",
+        dest="dataset",
```
Comment (daniel-j-h) [inline visiont/scripts/__main__.py:32]: We don't need `dest` here and below. With `--dataset` the arg will get stored in `args.dataset` already :hugs:

Comment (daniel-j-h) [reply visiont/scripts/__main__.py:32]: But also: the `--cmd` flags need `required=True` if you want them to be required (like dataset e.g.) otherwise `args.dataset` can be `None`

File: visiont/scripts/train.py [hunk 6]
```diff
@@ -89,27 +100,50 @@ def main(args):
...
-        Resize(1024),
-        RandomCrop(224),
-        RandomHorizontalFlip(p=0.5),
-        RandomApply([ColorJitter(brightness=0.8, contrast=0.8, saturation=0.8, hue=0.2)], p=0.8),
-        RandomApply([GaussianBlur((3, 3), (1.5, 1.5))], p=0.1),
-        RandomGrayscale(p=0.2),
-        ToTensor(),
-        Normalize(mean=mean, std=std),
-        ToPatches(16)])
+    # mean, std = [0.485, 0.456, 0.406], [0.229, 0.224, 0.225]
```
Comment (daniel-j-h) [inline visiont/scripts/train.py:105]: Why no `Normalize` anymore? We now feed in un-normalized images? Do you plan on getting the Open Images stats then?

File: visiont/models.py (hunk 4 above)
Comment (sandhawalia) [reply visiont/models.py:44]: `T`=Time dimension analogy from language modeling.

File: visiont/scripts/train.py (hunk 6 above)
Comment (sandhawalia) [reply visiont/scripts/train.py:105]: Token representations in transformers are not whitened `(mean=0, std=1)` not sure why we need them here.

**Comment by sandhawalia** [comment]
Hmm i'm still seeing a collapse 🙈 

```
visiont val --dataset /nas/3rd_party/openimagesV6/validation --weights /nas/team-space/experiments/vision-t/1-11-2010-db23f55/vt-010.pth
```

File: visiont/models.py (hunk 4 above)
Comment (sandhawalia) [reply visiont/models.py:44]: `D` == model dimension

File: visiont/scripts/train.py (hunk 6 above)
Comment (daniel-j-h) [reply visiont/scripts/train.py:105]: I don't follow; the token gets only cat-ed to the patch projections here, no?

https://github.com/daniel-j-h/visiont-pytorch/blob/4d9c4213aea9be291bd4f1b7708e2f103a639c52/visiont/models.py#L51
